### PR TITLE
Implement Testerina Code coverage for Multiple modules

### LINKED ISSUE #25791: [Testerina] [Code Coverage Report] Generating report using source files from other modules
**Description:**
The current implementation for Code-Coverage does not generate a coverage report when the test cases are kept in a separate module.

**Steps to reproduce:**
1. Create a ballerina project with 2 modules
2. Add some functions to the first module and write test cases for the functions in the second module
3. Run the tests using the code-coverage flag

**Affected Versions:**
slp3

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by Ibaqu**
## Purpose
Implement code coverage for test cases that cover source files from separate modules.

Fixes #25791

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/CoverageReport.java [hunk 1]
```diff
@@ -68,6 +69,9 @@ public CoverageReport(Path sourceJarPath, Path targetDirPath, String orgName, St
         this.classesDirectory = projectDir.resolve(TesterinaConstants.BIN_DIR);
         this.executionDataFile = projectDir.resolve(TesterinaConstants.EXEC_FILE_NAME);
         this.execFileLoader = new ExecFileLoader();
+
+        this.jarCache = targetDirPath.resolve("caches/jar_cache/").resolve(orgName);
```
Comment (azinneera) [inline misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/CoverageReport.java:73]: Lets resolve each child directory. i.e. `targetDirPath.resolve("caches").resolve("jar_cache").resolve(orgName)`

Comment (azinneera) [reply misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/CoverageReport.java:73]: Applies to L74 as well.

File: misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/ModuleCoverage.java [hunk 2]
```diff
@@ -1,3 +1,5 @@
+
```
Comment (azinneera) [inline misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/ModuleCoverage.java:1]: unnecessary new lines

File: misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/CoverageReport.java (hunk 1 above)
Comment (Ibaqu) [reply misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/CoverageReport.java]: Changed to resolve each child directory

File: misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/ModuleCoverage.java (hunk 2 above)
Comment (Ibaqu) [reply misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/ModuleCoverage.java]: Removed new lines

File: misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/CoverageReport.java [hunk 3]
```diff
@@ -76,60 +80,94 @@ public CoverageReport(Path sourceJarPath, Path targetDirPath, String orgName, St
      * @throws IOException when file operations are failed
      */
     public void generateReport() throws IOException {
-        try {
-            CodeCoverageUtils.unzipCompiledSource(sourceJarPath, projectDir, orgName, moduleName, version);
-        } catch (NoSuchFileException e) {
+
+        // Obtain a path list of all the .jar files generated
+        List<Path> pathList;
+        try (Stream<Path> walk = Files.walk(this.jarCache, 5)) {
```
Comment (Dilhasha) [inline misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/CoverageReport.java:86]: Just for my knowledge, how did we come up with this value as 5 for maxDepth?
Shall we define it as a constant?

File: misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/CoverageReport.java [hunk 4]
```diff
@@ -76,60 +80,94 @@ public CoverageReport(Path sourceJarPath, Path targetDirPath, String orgName, St
...
         analyzer.analyzeAll(classesDirectory.toFile());
         return coverageBuilder.getBundle(title);
     }
 
     private void createReport(final IBundleCoverage bundleCoverage) {
-        boolean containsSourceFiles;
+        boolean containsSourceFiles = true;
 
         for (IPackageCoverage packageCoverage : bundleCoverage.getPackages()) {
+            // I havent tested the behaviour of single files
```
Comment (Dilhasha) [inline misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/CoverageReport.java:125]: Is this a TODO?

File: misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/CoverageReport.java [hunk 5]
```diff
@@ -76,60 +80,94 @@ public CoverageReport(Path sourceJarPath, Path targetDirPath, String orgName, St
...
                         }
-                        ModuleCoverage.getInstance().addSourceFileCoverage(moduleName, sourceFileCoverage.getName(),
-                                coveredLines, missedLines);
+
+                        // Only add the source files that belong to the same module
+                        if (sourceFileModule.equals(moduleName)) {
+                            ModuleCoverage.getInstance().addSourceFileCoverage(sourceFileModule,
+                                    sourceFileCoverage.getName(), coveredLines, missedLines);
+                        } else {
+                            String jsonCachePath = this.jsonCache.toString() + "/"
```
Comment (Dilhasha) [inline misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/CoverageReport.java:162]: The module name can include "." as well right? 
Let's see we have a project with 2 modules as "module_test" & "module.test", will this have an impact?

File: misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/CoverageReport.java (hunk 4 above)
Comment (Ibaqu) [reply misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/CoverageReport.java:125]: Code coverage is actually not supported for single files. I checked. I will remove this TODO comment

File: misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/CoverageReport.java (hunk 5 above)
Comment (Ibaqu) [reply misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/CoverageReport.java:162]: Yes. This certainly is a problem. Although it doesn't cause any impact to the overall functionality, it still creates unnecessary cache files when a module has '_' in its name. 

I will fix this to only consider the version when replacing '_' with '.'

Comment (Ibaqu) [reply misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/CoverageReport.java]: I have added a function that deals with extracting the version and replacing the `_` with `.` 
This should only apply to the version and the module name will be spared.

File: misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/CoverageReport.java (hunk 3 above)
Comment (Ibaqu) [reply misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/CoverageReport.java]: Defined the Depth as a constant in TesterinaConstants

**Review by Dilhasha — APPROVED** [review]

**Pull request merged**